src/main/java/org/veupathdb/service/eda/ss/model/reducer/StreamIntersectMerger.java

```diff
@@ -1,92 +1,125 @@
+ * all input streams are sorted.
  */
 public class StreamIntersectMerger implements Iterator<Long> {
-  private final PeekableIterator[] streams;
-  private final int[] nextStream;
-  private Long currentElement;
-  private int currentStreamIdx;
-  private PeekableIterator currentStream;
+  private final PeekableIterator[] peekableIdIndexStreams;
+  private Long nextOutputIndex;
+  private PeekableIterator currentIdIndexStream;
   private boolean hasStarted;
+  private RingLinkedList streamRing;
 
   /**
    * @param sortedStreams Collection of sorted streams to merge by intersection.
    */
   public StreamIntersectMerger(List<Iterator<Long>> sortedStreams) {
-    this.streams = sortedStreams.stream()
+    // If no input streams are provided, it should act as an "empty" Iterator.
+    // If any input stream is "empty", should act as an "empty" Iterator since we are intersecting Iterators.
+    if (sortedStreams.isEmpty() || !sortedStreams.stream().allMatch(Iterator::hasNext)) {
+      hasStarted = true;
+      peekableIdIndexStreams = null;
+      nextOutputIndex = null;
+      return;
+    }
+    // Convert iterators into peekable iterators.
+    this.peekableIdIndexStreams = sortedStreams.stream()
         .map(PeekableIterator::new)
         .toArray(PeekableIterator[]::new);
-    this.currentStream = Arrays.stream(streams)
-        .max(Comparator.comparing(iter -> iter.peek()))
-        .get();
-    this.currentStreamIdx = IntStream.range(0, streams.length)
-        .filter(i -> streams[i] == currentStream)
-        .findFirst()
-        .getAsInt();
-    // Cache a pointer to the nextStream to avoid excessive modulo operations.
-    this.nextStream = IntStream.range(0, streams.length)
-        .map(i -> (i + 1) % streams.length)
-        .toArray();
+    this.streamRing = new RingLinkedList(peekableIdIndexStreams);
+    this.currentIdIndexStream = streamRing.cursor.iterator;
     hasStarted = false;
   }
 
   @Override
   public boolean hasNext() {
     if (!hasStarted) {
-      setCurrentElement();
+      findNextMatchingIdIndex();
       hasStarted = true;
     }
-    if (currentElement == null) {
-      return false;
-    } else {
-      return true;
-    }
+    return nextOutputIndex != null;
   }
 
   @Override
   public Long next() {
     if (!hasStarted) {
-      setCurrentElement();
+      findNextMatchingIdIndex();
       hasStarted = true;
     }
-    Long next = currentElement;
-    currentStream.next();
-    setCurrentElement();
+    Long next = nextOutputIndex;
+    currentIdIndexStream.next();
+    findNextMatchingIdIndex();
     return next;
   }
 
+  private void findNextMatchingIdIndex() {
+    boolean matchFound = false;
+    while (!matchFound && currentIdIndexStream.hasNext()) {
+      Long candidateIdIndex = currentIdIndexStream.peek();
+      // Iterate through each stream in the stream ring.
+      for (int i = 0; i < peekableIdIndexStreams.length; i++) {
```

Comment:
while cosmetic, i think it might make sense for RingLinkedList to have a `length` instance variable, and refer to it in this method instead of `peekableIdIndexStreams.length`

Reply:
Sure, fixed